src/Compilers/CSharp/Test/Emit2/Semantics/CollectionLiteralTests.cs

```diff
@@ -8655,6 +8655,219 @@ static void F(MyCollection<int> c)
+                using System.Runtime.CompilerServices;
+                [CollectionBuilder(typeof(MyCollectionBuilder), nameof(MyCollectionBuilder.Create))]
+                class MyCollection<T>
+                {
+                    internal IEnumerator<T> GetEnumerator() => default;
+                    public static MyCollection<T> F() => [];
+                }
+                class MyCollectionBuilder
+                {
+                    public static MyCollection<T> Create<T>(ReadOnlySpan<T> items) => default;
+                }
+                class Program
+                {
+                    static void Main()
+                    {
+                        MyCollection<int> c = [];
+                    }
+                }
+                """;
+            var comp = CreateCompilation(new[] { source, CollectionBuilderAttributeDefinition }, targetFramework: TargetFramework.Net70);
+            comp.VerifyEmitDiagnostics(
+                // 0.cs(8,42): error CS9184: 'MyCollection<T>' has a CollectionBuilderAttribute but no element type.
+                //     public static MyCollection<T> F() => [];
+                Diagnostic(ErrorCode.ERR_CollectionBuilderNoElementType, "[]").WithArguments("MyCollection<T>").WithLocation(8, 42),
+                // 0.cs(18,31): error CS9184: 'MyCollection<int>' has a CollectionBuilderAttribute but no element type.
+                //         MyCollection<int> c = [];
+                Diagnostic(ErrorCode.ERR_CollectionBuilderNoElementType, "[]").WithArguments("MyCollection<int>").WithLocation(18, 31));
+        }
+
+        [InlineData("", "", false)]
+        [InlineData("", "", true)]
+        [InlineData("scoped", "", false)]
+        [InlineData("scoped", "", true)]
+        [InlineData("scoped", "scoped", false)]
+        [InlineData("scoped", "scoped", true)]
+        [ConditionalTheory(typeof(CoreClrOnly))]
+        public void CollectionBuilder_Scoped(string constructorParameterModifier, string builderParameterModifier, bool useCompilationReference)
+        {
+            string sourceA = $$"""
+                using System;
+                using System.Collections.Generic;
+                using System.Runtime.CompilerServices;
+                [CollectionBuilder(typeof(MyCollectionBuilder), nameof(MyCollectionBuilder.Create))]
+                public ref struct MyCollection<T>
+                {
+                    private readonly List<T> _list;
+                    public MyCollection({{constructorParameterModifier}} ReadOnlySpan<T> items)
+                    {
+                        _list = new List<T>(items.ToArray());
+                    }
+                    public IEnumerator<T> GetEnumerator() => _list.GetEnumerator();
+                }
+                public class MyCollectionBuilder
+                {
+                    public static MyCollection<T> Create<T>({{builderParameterModifier}} ReadOnlySpan<T> items) => new(items);
```

Comment:
do we have a test which is attempting to return the resulting ref-like collection? Not blocking this PR.